toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -165,91 +65,210 @@ static auto GetBinaryOp(TokenKind kind) -> Semantics::BinaryOperator::Op {
+            << subtree_size;
+        auto name_node = node_stack.back().parse_node;
+        sem_ir.push_back(semantics_.nodes_.Store(Semantics::SetName(
+            name_node, parse_tree().GetNodeText(name_node), function_id)));
+        node_stack.pop_back();
+
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir)));
+        break;
+      }
+      case ParseNodeKind::FileEnd(): {
+        CARBON_CHECK(subtree_size == 1);
+        ++it;
+        CARBON_CHECK(it == range.end())
+            << "FileEnd should always be last, found "
+            << parse_tree().node_kind(*it);
+
+        for (const auto& entry : node_stack) {
+          semantics_.root_block_.append(entry.sem_ir.begin(),
+                                        entry.sem_ir.end());
+        }
+        return;
+      }
+      case ParseNodeKind::InfixOperator(): {
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+
+        subtree_size -=
+            parse_tree().node_subtree_size(node_stack.back().parse_node);
+        sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                      node_stack.back().sem_ir.end());
+        auto rhs_id = *node_stack.back().result_id;
+        node_stack.pop_back();
+
+        subtree_size -=
+            parse_tree().node_subtree_size(node_stack.back().parse_node);
+        sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                      node_stack.back().sem_ir.end());
+        auto lhs_id = *node_stack.back().result_id;
+        node_stack.pop_back();
+
+        CARBON_CHECK(subtree_size == 1);
+
+        // Figure out the operator for the token.
+        auto token = parse_tree().node_token(parse_node);
+        auto token_kind = tokens_->GetKind(token);
+        auto op = GetBinaryOp(token_kind);
+
+        auto literal_id = next_id();
+        sem_ir.push_back(semantics_.nodes_.Store(Semantics::BinaryOperator(
+            parse_node, literal_id, op, lhs_id, rhs_id)));
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir), literal_id));
+        break;
+      }
+      case ParseNodeKind::Literal(): {
+        RequireNodeEmpty(parse_node, parse_kind, subtree_size);
+        auto literal_id = next_id();
+
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+        auto token = parse_tree().node_token(parse_node);
+        switch (auto token_kind = tokens_->GetKind(token)) {
+          case TokenKind::IntegerLiteral(): {
+            sem_ir.push_back(semantics_.nodes_.Store(Semantics::IntegerLiteral(
+                parse_node, literal_id, tokens_->GetIntegerLiteral(token))));
+            break;
+          }
+          default:
+            CARBON_FATAL() << "Unhandled kind: " << token_kind.Name();
+        }
+        // TODO: This should transform into a usable parameter list. For now
+        // it's unused and only stored so that node counts match.
```

Comment:
I don't follow this TODO; should it be here? The copy in `case ParseNodeKind::ParameterList` below makes sense to me.

Reply:
Oops, copy pasta, removed